src/libraries/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.AddSub.cs

```diff
@@ -100,53 +173,103 @@ public static void Subtract(ReadOnlySpan<uint> left, uint right, Span<uint> bits
+                    i++;
+                    carry >>= 32;
+                    if (carry == 0) break;
+                }
+
+                if (i < upperBound)
+                {
+                    CopyTail(left, bits, unchecked((int)i));
+                }
+            }
+
         }
 
         public static void Subtract(ReadOnlySpan<uint> left, ReadOnlySpan<uint> right, Span<uint> bits)
         {
+            Debug.Assert(right.Length >= 1);
             Debug.Assert(left.Length >= right.Length);
             Debug.Assert(Compare(left, right) >= 0);
             Debug.Assert(bits.Length == left.Length);
 
-            int i = 0;
-            long carry = 0L;
-
             // Switching to managed references helps eliminating
             // index bounds check...
-            ref uint leftPtr = ref MemoryMarshal.GetReference(left);
             ref uint resultPtr = ref MemoryMarshal.GetReference(bits);
+            ref uint rightPtr = ref MemoryMarshal.GetReference(right);
+            ref uint leftPtr = ref MemoryMarshal.GetReference(left);
 
-            // Executes the "grammar-school" algorithm for computing z = a - b.
-            // While calculating z_i = a_i - b_i we take care of overflow:
-            // Since a_i - b_i doesn't need any additional bit, our carry c
-            // has always the value -1 or 0; hence, we're safe here.
+            nint i = 0;
+            nint upperBound = right.Length;
+            long carry = 0;
 
-            for ( ; i < right.Length; i++)
+            // Executes the "grammar-school" algorithm for computing z = a + b.
+            // While calculating z_i = a_i + b_i we take care of overflow:
+            // Since a_i + b_i + c <= 2(2^32 - 1) + 1 = 2^33 - 1, our carry c
+            // has always the value 1 or 0; hence, we're safe here.
+
+            do
             {
-                long digit = (Unsafe.Add(ref leftPtr, i) + carry) - right[i];
-                Unsafe.Add(ref resultPtr, i) = unchecked((uint)digit);
-                carry = digit >> 32;
+                carry += Unsafe.Add(ref leftPtr, i);
+                carry -= Unsafe.Add(ref rightPtr, i);
+                Unsafe.Add(ref resultPtr, i) = unchecked((uint)carry);
+                carry >>= 32;
+                i++;
+            } while (i < upperBound);
+            upperBound = left.Length;
+
+            if (upperBound <= CopyToThreshold)
+            {
+                for ( ; i < upperBound; i++)
+                {
+                    carry += Unsafe.Add(ref leftPtr, i);
+                    Unsafe.Add(ref resultPtr, i) = unchecked((uint)carry);
+                    carry >>= 32;
+                }
             }
-            for ( ; i < left.Length; i++)
+            else
             {
-                long digit = left[i] + carry;
-                Unsafe.Add(ref resultPtr, i) = (uint)digit;
-                carry = digit >> 32;
+                for ( ; carry != 0 && i < upperBound; i++)
```

Comment:
is there any particular reason why in other loops you have added a break for `carry == 0` while here you moved the condition here?